File: hw5/task3.py
from blackbox import BlackBox
import sys
from pyspark import SparkContext
import os
import time
import random

if __name__ == "__main__":

    inputs = sys.argv[1]
    stream = int(sys.argv[2])
    asks = int(sys.argv[3])
    outputs = sys.argv[4]

    start_time = time.time()
    #sc.setLogLevel("ERROR")
    random.seed(553)
    bx = BlackBox()
    result_header = "seqnum,0_id,20_id,40_id,60_id,80_id\n"
    user_list = []
    total = 0
    size = 100
    #result_h = ""
    for i in range(asks):
        # Simulate streaming users from the BlackBox
        users_stream = bx.ask(inputs, stream)
        for user in users_stream:
            total += 1
            if len(user_list) < size:
                user_list.append(user)

            else:
                if random.random() < size / total:
                    user_list[random.randint(0, size - 1)] = user
            if total % size == 0:
                elem = [str(total), str(user_list[0]), str(user_list[20]), str(user_list[40]), str(user_list[60]),
                        str(user_list[80])]
                result_header += ','.join(elem) + '\n'
    with open(outputs, 'w') as file:
        file.write(result_header)
    end_time = time.time()
    execution = end_time - start_time
    print(f"Duration: {execution}")

File: hw5/task3.scala
import scala.io.Source
import java.io._
import scala.util.Random
import scala.collection.mutable.ArrayBuffer



object task3 {
  val random = new Random(553) // A single Random instance for use throughout

  def main(args: Array[String]): Unit = {

    val startTime = System.currentTimeMillis()
    val inputs = args(0)
    val streamSize = args(1).toInt
    val asks = args(2).toInt
    val outputs = args(3)

    val bx = new Blackbox()
    var userList = ArrayBuffer.empty[String]
    var total = 0
    val size = 100
    var resultHeader = "seqnum,0_id,20_id,40_id,60_id,80_id\n"

    for (_ <- 1 to asks) {
      val usersStream = bx.ask(inputs, streamSize)
      for  (user <- usersStream) {
        total += 1
        if (userList.length < size) {
          userList += user
        } else {
          if (Random.nextFloat() < 100.0 / total) {
            userList(Random.nextInt(100)) = user
          }
        }
      }
      if (total % 100 == 0) {
        val seqNum = total
        val elements = Seq(seqNum.toString, userList(0), userList(20), userList(40), userList(60), userList(80))
        resultHeader += elements.mkString(",") + "\n"
      }
    }

    // Write the results to a file
    val writer = new PrintWriter(new File(outputs))
    writer.write(resultHeader)
    writer.close()

    val endTime = System.currentTimeMillis()
    println(f"Duration: ${(endTime - startTime) / 1000} seconds")
  }
}
